# Análisis



In [ ]:
#config PostgreSQL
!apt install postgresql postgresql-contrib &> log
!service postgresql start
!sudo -u postgres psql -c "CREATE USER root WITH SUPERUSER"

%load_ext sql
%config SqlMagic.feedback = False
%config SqlMagic.autopandas = True

In [ ]:
#config ngrok
%sql postgresql+psycopg2://postgres:36204@5.tcp.eu.ngrok.io:16267/postgres

### 1. ETL con información relativa a los empleados de la compañía a 31-12-2002

 Identificamos primero a los empleados que estaban activos a finales de 2002. Luego se une a las otras tablas para obtener los salarios, número de empleado, rol, género, edad, antigüedad y departamentos actuales de esos empleados. Se ordena de mayor a menor en cuanto a salarios, simplemnte para visualizarlo mejor y tener un orden.

In [ ]:
%%sql
WITH active_employees AS (
    -- Identificamos los empleados activos el 31-12-2002
    SELECT emp_no
    FROM dept_emp
    WHERE from_date <= '2002-12-31' AND to_date >= '2002-12-31'
),

employee_details AS (
    -- Obtenemos los detalles relevantes de los empleados
    SELECT
        e.emp_no,
        s.amount AS salarios,
        t.title AS rol,
        e.gender AS genero,
        EXTRACT(YEAR FROM age('2002-12-31', e.birth_date))::INT AS edad,
        EXTRACT(YEAR FROM age('2002-12-31', e.hire_date))::INT AS antiguedad,
        de.dept_no
    FROM active_employees ae
    JOIN employee e ON ae.emp_no = e.emp_no
    LEFT JOIN salary s ON ae.emp_no = s.emp_no AND s.from_date <= '2002-12-31' AND s.to_date > '2002-12-31'
    LEFT JOIN title t ON ae.emp_no = t.emp_no AND t.from_date <= '2002-12-31' AND t.to_date > '2002-12-31'
    LEFT JOIN current_dept_emp de ON ae.emp_no = de.emp_no
)

SELECT
    ed.emp_no,
    ed.salarios,
    ed.rol,
    ed.genero,
    ed.edad,
    ed.antiguedad,
    d.dept_name AS departamento
FROM employee_details ed
LEFT JOIN department d ON ed.dept_no = d.dept_no
ORDER BY ed.salarios DESC;

 * postgresql+psycopg2://postgres:***@5.tcp.eu.ngrok.io:16267/postgres


emp_no  salarios               rol genero  edad  antiguedad  \
0     10897    136004      Senior Staff      F    44          16   
1     10539    126614      Senior Staff      M    43          12   
2     10548    126034      Senior Staff      M    50          16   
3     10955    125713      Senior Staff      F    43          16   
4     10532    122653      Senior Staff      F    43          16   
..      ...       ...               ...    ...   ...         ...   
797   10715     42510          Engineer      M    43           7   
798   10997     42260  Technique Leader      F    43           8   
799   10022     41348          Engineer      M    50           7   
800   10766     41112             Staff      M    38           6   
801   10823     40063             Staff      M    46           7   

        departamento  
0              Sales  
1              Sales  
2          Marketing  
3              Sales  
4              Sales  
..               ...  
797       Production  
798       Production  
799      Development  
800  Human Resources  
801  Human Resources  

[802 rows x 7 columns]

Lo agrupamos por rol, para que lo visualizemos mejor:

In [ ]:
%%sql
WITH active_employees AS (
    -- Identificamos los empleados activos el 31-12-2002
    SELECT emp_no
    FROM dept_emp
    WHERE from_date <= '2002-12-31' AND to_date >= '2002-12-31'
),

employee_details AS (
    -- Obtenemos los detalles relevantes de los empleados
    SELECT
        e.emp_no,
        s.amount AS salarios,
        t.title AS rol,
        e.gender AS genero,
        EXTRACT(YEAR FROM age('2002-12-31', e.birth_date))::INT AS edad,
        EXTRACT(YEAR FROM age('2002-12-31', e.hire_date))::INT AS antiguedad,
        de.dept_no
    FROM active_employees ae
    JOIN employee e ON ae.emp_no = e.emp_no
    LEFT JOIN salary s ON ae.emp_no = s.emp_no AND s.from_date <= '2002-12-31' AND s.to_date > '2002-12-31'
    LEFT JOIN title t ON ae.emp_no = t.emp_no AND t.from_date <= '2002-12-31' AND t.to_date > '2002-12-31'
    LEFT JOIN current_dept_emp de ON ae.emp_no = de.emp_no
)

SELECT
    ed.rol,
    ROUND(AVG(ed.salarios)) AS salario_promedio,
    ROUND(AVG(ed.edad)) AS edad_promedio,
    ROUND(AVG(ed.antiguedad)) AS antiguedad_promedio,
    d.dept_name AS departamento,
    COUNT(ed.emp_no) AS numero_empleados
FROM employee_details ed
LEFT JOIN department d ON ed.dept_no = d.dept_no
GROUP BY ed.rol, d.dept_name
ORDER BY salario_promedio DESC;


 * postgresql+psycopg2://postgres:***@5.tcp.eu.ngrok.io:16267/postgres


rol salario_promedio edad_promedio antiguedad_promedio  \
0         Senior Staff            88946            43                  13   
1         Senior Staff            82749            45                  13   
2         Senior Staff            80351            43                  12   
3     Technique Leader            79262            43                  13   
4                Staff            77826            44                  11   
5                Staff            77792            44                  12   
6      Senior Engineer            73799            44                  13   
7      Senior Engineer            71947            44                  13   
8     Technique Leader            71824            45                  12   
9         Senior Staff            71053            43                  13   
10     Senior Engineer            69771            44                  13   
11               Staff            69733            44                  10   
12        Senior Staff            68544            44                  13   
13    Technique Leader            67505            43                  11   
14     Senior Engineer            67319            43                  13   
15        Senior Staff            67129            44                  13   
16     Senior Engineer            66276            45                  14   
17        Senior Staff            65171            45                  10   
18            Engineer            64815            46                  11   
19  Assistant Engineer            64432            44                   9   
20        Senior Staff            62641            45                  14   
21               Staff            61569            45                   9   
22            Engineer            60991            44                  11   
23            Engineer            60903            44                  11   
24               Staff            60255            44                  11   
25    Technique Leader            56716            44                  12   
26  Assistant Engineer            55541            43                  16   
27  Assistant Engineer            55411            42                  10   
28            Engineer            55385            38                   8   
29            Engineer            55138            42                  12   
30               Staff            51982            43                   9   
31               Staff            50549            43                  11   

          departamento  numero_empleados  
0                Sales                97  
1            Marketing                28  
2              Finance                27  
3             Research                 3  
4                Sales                33  
5            Marketing                 6  
6          Development               118  
7           Production               115  
8          Development                22  
9             Research                27  
10            Research                11  
11             Finance                16  
12    Customer Service                38  
13          Production                19  
14  Quality Management                31  
15     Human Resources                35  
16    Customer Service                 7  
17         Development                 3  
18  Quality Management                 8  
19          Production                 6  
20          Production                 3  
21            Research                 8  
22         Development                56  
23          Production                40  
24          Production                 3  
25  Quality Management                 7  
26         Development                 4  
27  Quality Management                 2  
28    Customer Service                 1  
29            Research                 2  
30     Human Resources                11  
31    Customer Service                15

### 2. ¿Se pordría asumir que el ambiente y las condiciones laborables eran buenos para el empleado? ¿Cuánto gastaba la empresa en salarios anualmente a fecha 31-12-2002? ¿Cuánto estimas que era la facturación anual de la empresa?

Para saber si hay un buen ambiente y condiciones buenas para el empleado en la empresa UAX Tech a fecha de 31-12-2002, podemos ver:


- La media de Diferencia de edad: Si hay mucha diferéncia puede perjudicar el ambiente laboral, puede haber diferentes entendimientos entre los empleados.
- Diferencia de genero: Si encontrameos que no hay igualdad en cuanto a género de empleados, es posible que perjudique al ambiente laboral.
- Antigüedad de los Empleados: Si muchos empleados han estado en la empresa por mucho tiempo, esto podría indicar condiciones de trabajo favorables.
- Consistencia en el Salario: Si el salario de un empleado ha aumentado de manera constante a lo largo del tiempo, esto podría ser un indicador de reconocimiento y recompensa por su trabajo.
- La media de los salarios para cada tipo de puesto.

**La media de Diferencia de edad**: Podemos ver que la diferencia de media en edad a fecha de 31-12-2002 es de 3 años.

In [ ]:
%%sql
WITH employee_age AS (
    SELECT
        emp_no,
        EXTRACT(YEAR FROM age('2002-12-31', birth_date))::INT AS edad
    FROM employee
)

, average_age AS (
    SELECT AVG(edad) AS edad_promedio
    FROM employee_age
)

SELECT
    ROUND(
    AVG(ABS(edad - (SELECT edad_promedio FROM average_age)))) AS diferencia_media_edad
FROM employee_age;



 * postgresql+psycopg2://postgres:***@5.tcp.eu.ngrok.io:16267/postgres


diferencia_media_edad
0                     3

**Diferencia de genero**: Hcemos la consulta deonde primero determina el número de empleados activos de cada género en 31-12-2002 y luego se calcula el porcentaje de cada género y la diferencia con el 100%, con esto deducimos fácilmente la diferencia entre el porcentaje de empleados masculinos y femeninos.

In [ ]:
%%sql
-- Identificamos a todos los empleados activos a fecha de 31-12-2002
WITH active_employees AS (
    SELECT e.emp_no, e.gender
    FROM employee e
    JOIN dept_emp de ON e.emp_no = de.emp_no
    WHERE de.from_date <= '2002-12-31' AND de.to_date >= '2002-12-31'
)

-- Contamos el número de empleados masculinos y femeninos y calculamos la diferencia de porcentajes
, gender_counts AS (
    SELECT
        gender,
        ROUND(COUNT(emp_no) * 100.0 / (SELECT COUNT(emp_no) FROM active_employees), 2) AS percentage
    FROM active_employees
    GROUP BY gender
)

SELECT
    gender,
    percentage
FROM gender_counts;



 * postgresql+psycopg2://postgres:***@5.tcp.eu.ngrok.io:16267/postgres


gender percentage
0      M      60.10
1      F      39.90

**Antigüedad de los Empleados**: A fecha de 31-12-2002, calculamos la antiguedad promedio, la mínima y máxima.

In [ ]:
%%sql
WITH Antiguedad AS (
    SELECT
        emp_no,
        EXTRACT(YEAR FROM age('2002-12-31', hire_date))::INT AS antiguedad
    FROM employee
    WHERE '2002-12-31' BETWEEN hire_date AND COALESCE(
        (SELECT to_date
         FROM dept_emp
         WHERE emp_no = employee.emp_no
         ORDER BY to_date DESC LIMIT 1), '2002-12-31')
)
SELECT
    ROUND(AVG(antiguedad)) AS antiguedad_promedio,
    MIN(antiguedad) AS antiguedad_minima,
    MAX(antiguedad) AS antiguedad_maxima
FROM Antiguedad;


 * postgresql+psycopg2://postgres:***@5.tcp.eu.ngrok.io:16267/postgres


antiguedad_promedio  antiguedad_minima  antiguedad_maxima
0                  12                  3                 17

**Consistencia en el Salario**: Calculamos el número total de empleados, el número de empleados con aumentos consistentes en el salario y la proporción (en porcentaje) de empleados con aumentos consistentes en relación con el total, a fecha de 31-12-2002

In [ ]:
%%sql
WITH salary_differences AS (
    -- Calcula la diferencia entre salarios consecutivos para cada empleado
    SELECT
        emp_no,
        amount - LAG(amount) OVER (PARTITION BY emp_no ORDER BY from_date) AS diff,
        from_date,
        to_date
    FROM salary
    WHERE to_date <= '2002-12-31'
),

consistent_salaries AS (
    -- Identifica empleados que han tenido aumentos consistentes en su salario
    SELECT
        emp_no
    FROM salary_differences
    GROUP BY emp_no
    HAVING MIN(diff) > 0 -- Todos los cambios en el salario deben ser positivos
)

-- Calcular la proporción de empleados con aumentos consistentes en relación con el total de empleados
SELECT
    COUNT(DISTINCT emp_no) AS total_empleados,
    (SELECT COUNT(emp_no) FROM consistent_salaries) AS empleados_con_aumentos_consistentes,
    (SELECT COUNT(emp_no) FROM consistent_salaries)::FLOAT / COUNT(DISTINCT emp_no) * 100 AS porcentaje_con_aumentos_consistentes
FROM employee
WHERE '2002-12-31' >= hire_date;


 * postgresql+psycopg2://postgres:***@5.tcp.eu.ngrok.io:16267/postgres


total_empleados  empleados_con_aumentos_consistentes  \
0             1000                                  468   

   porcentaje_con_aumentos_consistentes  
0                                  46.8

**La media de los salarios para cada tipo de puesto**:

In [ ]:
%%sql
SELECT
    t.title AS rol,
    ROUND(AVG(s.amount))AS salario_promedio
FROM
    title t
    JOIN salary s ON t.emp_no = s.emp_no
WHERE
    t.from_date <= '2002-12-31' AND ('2002-12-31' <= t.to_date OR t.to_date IS NULL)
    AND s.from_date <= '2002-12-31' AND '2002-12-31' <= s.to_date
GROUP BY t.title
ORDER BY salario_promedio DESC;

 * postgresql+psycopg2://postgres:***@5.tcp.eu.ngrok.io:16267/postgres


rol salario_promedio
0        Senior Staff            78955
1     Senior Engineer            71988
2    Technique Leader            68579
3               Staff            66892
4            Engineer            61082
5  Assistant Engineer            59965

**Conclusión**: Podemos afirmar que a fecha de 31-12-2002 la empresa UAX Tech disponde de un buen ambiente y condiciones laborables, ya que:

- En cuanto a **la media de Diferencia de edad:** Los empleados tienen una diferencia de edad promedio de 3 años, indicando un buen equilibrio de edad que favorece el entendimiento y energía del equipo.

- **Diferencia de genero**: A fecha de 31-12-2002 los empleados estan distribuïdos en 60,1% Hombres y 39,9% Mujeres, por lo tano sí existe una brecha de género en cuanto a núemro de empleados se refiere. Consideramos que no es una diférencia significatibamente alta para que perjudique el ambiente laboral.

- **Antigüedad de los Empleados**: Bajo la hipótesis de que el mercado laboral de EEUU a fecha de 31-12-2002 no esta en crisis. La antigüedad promedio es de 12 años, con una antigüedad máxima de 17 años y una mínima de 3 años. Indica que UAX Tech es visto como un lugar favorable para desarrollar carreras profesionales.

- **Consistencia en el Salario**: Casi la mitad (46,8%) de los empleados han tenido un aumento salarial hasta finales de 2002, lo que refleja condiciones salariales positivas.

- **La media de los salarios para cada tipo de puesto**:  Aunque se necesita una comparación con el mercado laboral, los salarios parecen ser altos y apropiados para la responsabilidad del puesto, lo que sugiere buenas condiciones para los empleados.

**Gasto salarial anual de la empresa a fecha de 31-12-2002**: Sumamos todos los salarios de los empleados que estuvieron activos durante ese año.

Muy imporante, hay que considerar que un empleado podría haber tenido más de un salario en el año 2002 debido a aumentos o cambios de puesto. Por eso mismo, considemos los periodos correctos al sumar los salarios:

In [ ]:
%%sql
WITH salaries_2002 AS (
    -- Extraemos los salarios de 2002, ajustando fechas si es necesario
    SELECT
        emp_no,
        amount,

        -- Si el salario comenzó antes de 2002, lo ajustamos al 1ro de enero

        CASE WHEN from_date < '2002-01-01' THEN '2002-01-01' ELSE from_date END AS adjusted_from_date,

        -- Si el salario finalizó después de 2002, lo ajustamos al 31 de diciembre

        CASE WHEN to_date > '2002-12-31' OR to_date = '9999-01-01' THEN '2002-12-31' ELSE to_date END AS adjusted_to_date
    FROM salary
    WHERE (from_date BETWEEN '2002-01-01' AND '2002-12-31')
       OR (to_date BETWEEN '2002-01-01' AND '2002-12-31')
       OR (from_date <= '2002-01-01' AND (to_date >= '2002-12-31' OR to_date = '9999-01-01'))
)

-- Calculamos el gasto salarial total basado en los días trabajados y el salario diario

SELECT
    ROUND(SUM(amount * (EXTRACT(DOY FROM adjusted_to_date) - EXTRACT(DOY FROM adjusted_from_date) + 1) / 365)) AS total_salario_2002
FROM salaries_2002;


 * postgresql+psycopg2://postgres:***@5.tcp.eu.ngrok.io:16267/postgres


total_salario_2002
0           57756882

**Facturación anual de la empresa**: En la base de datos, no se dispone de datos directos sobre la facturación anual. Aunque podemos estimar la facturación deL año 2002 basándonos en la información de que el gasto en salarios fue un 10% superior a la facturación de ese año (que fue la razón de quiebra).

In [ ]:
%%sql
WITH salaries_2002 AS (
    -- Extraemos los salarios de 2002, ajustando fechas si es necesario
    SELECT
        emp_no,
        amount,
        -- Si el salario comenzó antes de 2002, lo ajustamos al 1ro de enero
        CASE WHEN from_date < '2002-01-01' THEN '2002-01-01' ELSE from_date END AS adjusted_from_date,
        -- Si el salario finalizó después de 2002, lo ajustamos al 31 de diciembre
        CASE WHEN to_date > '2002-12-31' OR to_date = '9999-01-01' THEN '2002-12-31' ELSE to_date END AS adjusted_to_date
    FROM salary
    WHERE (from_date BETWEEN '2002-01-01' AND '2002-12-31')
       OR (to_date BETWEEN '2002-01-01' AND '2002-12-31')
       OR (from_date <= '2002-01-01' AND (to_date >= '2002-12-31' OR to_date = '9999-01-01'))
),

total_salaries_2002 AS (
    -- Calculamos el gasto salarial total basado en los días trabajados y el salario diario
    SELECT
        SUM(amount * (EXTRACT(DOY FROM adjusted_to_date) - EXTRACT(DOY FROM adjusted_from_date) + 1) / 365) AS total_salario
    FROM salaries_2002
)

SELECT
    ROUND(total_salario / 1.10, 2) AS facturacion_estimada_2002
FROM total_salaries_2002;



 * postgresql+psycopg2://postgres:***@5.tcp.eu.ngrok.io:16267/postgres


facturacion_estimada_2002
0               52506256.15

### 3. Análisis de la distribución salarial de la compañia en función de distintas variables como el género, edad, antigüedad o rol. ¿Se podría afirmar que existía brecha de género en la compañia?

Análisis de la distibución salarial en función del género, edad, antigüedad y rol:

- Análisis salarial por género: Comparar el salario promedio de hombres y mujeres.
- Análisis salarial por edad y género: Ver si existen diferencias salariales dentro de distintos grupos de edad.
- Análisis salarial por antigüedad y género: Comparar los salarios en función de cuánto tiempo ha estado un empleado en la empresa.
- Análisis salarial por rol y género: Analizar si hay diferencias de salario entre hombres y mujeres que desempeñan el mismo rol.

**Análisis salarial por género de la empresa**: Observamos que de media las mujeres cobran más que los hombres, unsa diferéncia positiva de 1555.67$.

In [ ]:
%%sql
SELECT
    e.gender,
    ROUND(AVG(s.amount), 2) AS avg_salary
FROM
    salary s
JOIN
    employee e ON s.emp_no = e.emp_no
GROUP BY
    e.gender;


 * postgresql+psycopg2://postgres:***@5.tcp.eu.ngrok.io:16267/postgres


gender avg_salary
0      M   63614.77
1      F   65170.44

Análisis salarial por edad y género: Podemos ver la diferencia entre los salarios promedio de hombres y mujeres por grupo de edad:
- Las mujeres cobran aproximadamente más en edades más avanzadas en comparación con los hombres.
- Los hombres cobran aproximadamente más en edades más jovenes.


In [ ]:
%%sql
WITH AgeGenderSalary AS (
    SELECT
        EXTRACT(YEAR FROM age('2002-12-31', e.birth_date)) AS age,
        e.gender,
        ROUND(AVG(s.amount),2) AS avg_salary
    FROM
        salary s
    JOIN
        employee e ON s.emp_no = e.emp_no
    WHERE
        s.from_date <= '2002-12-31' AND s.to_date > '2002-12-31'
    GROUP BY
        e.gender, age
)

SELECT
    a1.age,
    COALESCE(a1.avg_salary, 0) AS male_avg_salary,
    COALESCE(a2.avg_salary, 0) AS female_avg_salary,
    COALESCE(a1.avg_salary, 0) - COALESCE(a2.avg_salary, 0) AS salary_difference
FROM
    (SELECT age, avg_salary FROM AgeGenderSalary WHERE gender = 'M') a1
FULL JOIN
    (SELECT age, avg_salary FROM AgeGenderSalary WHERE gender = 'F') a2
ON
    a1.age = a2.age
ORDER BY
    a1.age DESC;


 * postgresql+psycopg2://postgres:***@5.tcp.eu.ngrok.io:16267/postgres


age male_avg_salary female_avg_salary salary_difference
0   50        69449.35          70324.74           -875.39
1   49        72595.44          69743.35           2852.09
2   48        69408.00          69336.62             71.38
3   47        72219.72          73754.05          -1534.33
4   46        68612.97          78258.18          -9645.21
5   45        66913.26          73033.70          -6120.44
6   44        69685.73          77133.65          -7447.92
7   43        69077.16          78124.83          -9047.67
8   42        68532.15          71443.00          -2910.85
9   41        73951.44          73687.69            263.75
10  40        79106.95          69509.67           9597.28
11  39        73573.15          67512.22           6060.93
12  38        71783.90          68218.04           3565.86
13  37        75502.00          60672.00          14830.00

Brecha salarial por rol y género: Vemos que por lo general las mujeres incuso siendo menos cobran más que los hombres. Esto nos indica que la empresa respeta y valora al género femenino sobre el masculino en cuanto a salario se refiere.

In [ ]:
%%sql
WITH GenderRoleSalary AS (
    -- Subconsulta que calcula el salario promedio para cada combinación de rol y género.
    SELECT
        e.gender,
        t.title,
        ROUND(AVG(s.amount),2) AS avg_salary,
        COUNT(e.emp_no) AS employee_count
    FROM
        salary s
    JOIN
        employee e ON s.emp_no = e.emp_no
    JOIN
        title t ON t.emp_no = e.emp_no
    WHERE
        (t.to_date = '9999-01-01' OR t.to_date > '2002-12-31')
    GROUP BY
        e.gender, t.title
)

SELECT
    grsM.title,
    grsM.avg_salary AS male_avg_salary,
    grsM.employee_count AS male_count,
    grsF.avg_salary AS female_avg_salary,
    grsF.employee_count AS female_count,
    COALESCE(grsM.avg_salary, 0) - COALESCE(grsF.avg_salary, 0) AS salary_difference
FROM
    (SELECT title, avg_salary, employee_count FROM GenderRoleSalary WHERE gender = 'M') grsM
LEFT JOIN
    (SELECT title, avg_salary, employee_count FROM GenderRoleSalary WHERE gender = 'F') grsF
ON
    grsM.title = grsF.title
ORDER BY
    salary_difference;


 * postgresql+psycopg2://postgres:***@5.tcp.eu.ngrok.io:16267/postgres


title male_avg_salary  male_count female_avg_salary  \
0  Assistant Engineer        51999.86          44          63591.58   
1        Senior Staff        67558.35        1839          72124.54   
2    Technique Leader        60549.04         268          63176.77   
3     Senior Engineer        61728.31        2019          62442.21   
4            Engineer        56561.85         416          56088.80   
5               Staff        64108.74         354          60597.94   

   female_count salary_difference  
0            19         -11591.72  
1          1242          -4566.19  
2           239          -2627.73  
3          1386           -713.90  
4           282            473.05  
5           174           3510.80

Realizamos la media de las columnas: male_avg_salary	male_count	female_avg_salary	female_count	salary_difference

In [ ]:
%%sql
WITH GenderRoleSalary AS (
    -- Subconsulta que calcula el salario promedio para cada combinación de rol y género.
    SELECT
        e.gender,
        t.title,
        ROUND(AVG(s.amount),2) AS avg_salary,
        COUNT(e.emp_no) AS employee_count
    FROM
        salary s
    JOIN
        employee e ON s.emp_no = e.emp_no
    JOIN
        title t ON t.emp_no = e.emp_no
    WHERE
        (t.to_date = '9999-01-01' OR t.to_date > '2002-12-31')
    GROUP BY
        e.gender, t.title
), GenderSalaryDifference AS (
    SELECT
        grsM.title,
        grsM.avg_salary AS male_avg_salary,
        grsM.employee_count AS male_count,
        grsF.avg_salary AS female_avg_salary,
        grsF.employee_count AS female_count,
        COALESCE(grsM.avg_salary, 0) - COALESCE(grsF.avg_salary, 0) AS salary_difference
    FROM
        (SELECT title, avg_salary, employee_count FROM GenderRoleSalary WHERE gender = 'M') grsM
    LEFT JOIN
        (SELECT title, avg_salary, employee_count FROM GenderRoleSalary WHERE gender = 'F') grsF
    ON
        grsM.title = grsF.title
)

SELECT
    ROUND(AVG(male_avg_salary), 2) AS avg_male_avg_salary,
    ROUND(AVG(male_count), 2) AS avg_male_count,
    ROUND(AVG(female_avg_salary), 2) AS avg_female_avg_salary,
    ROUND(AVG(female_count), 2) AS avg_female_count,
    ROUND(AVG(salary_difference), 2) AS avg_salary_difference
FROM
    GenderSalaryDifference;


 * postgresql+psycopg2://postgres:***@5.tcp.eu.ngrok.io:16267/postgres


avg_male_avg_salary avg_male_count avg_female_avg_salary avg_female_count  \
0            60417.69         823.33              63003.64           557.00   

  avg_salary_difference  
0              -2585.95

**Conclusión**: Existe una brecha salarial a favor del género femenino ya que:
- De media las mujeres cobran más que los hombres, una diferéncia positiva de 1555.67$ a favor del género femenino.

- En género femenino incuso siendo menos númeroso en todos los roles de la empresa, en los puestos de Assistant Engineer, Senior Staff, Technique Leader y Senior Engineer cobran más que los hombres.

- Las mujeres de media cobran 2585.95$ más que los hombres.


### 4. Web Scraping:


Extraemos los datos de la web builtin.com mediante BeautifulSoup

In [ ]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

titles = ["software-engineer", "senior-software-engineer", "qa-engineer", "systems-engineer", "senior-systems-engineer", "principal-software-engineer"]

salarios = []

for title in titles:
    url = f"https://builtin.com/salaries/dev-engineer/{title}"
    req = requests.get(url)
    soup = BeautifulSoup(req.text, 'html.parser')

    info = soup.find_all("div", class_='block block-bix-salary block-bix-salary-calculation-info')

    for item in info:
        nombre = item.find("h1", class_="box-title")
        total = item.find("div", class_="total").text

        salarios.append({
            "title": nombre,
            "salary": total
        })

df = pd.DataFrame(salarios)

df['title'] = df['title'].apply(lambda x: x.get_text().replace('[\[\]]*', '',True).replace('salary in US','').strip())
df['salary'] = df['salary'].apply(lambda x: x.replace('$', '').replace(',',''))
df['salary'] = df['salary'].astype(int)

df

title  salary
0            Software Engineer  135575
1     Senior Software Engineer  195208
2                  QA Engineer   98554
3             Systems Engineer  123232
4      Senior Systems Engineer  144377
5  Principal Software Engineer  169861

In [ ]:
# Cambio de nombre de los roles
title_mapping = {
    'Software Engineer': 'Engineer',
    'Senior Software Engineer': 'Senior Engineer',
    'QA Engineer': 'Assistant Engineer',
    'Systems Engineer': 'Staff',
    'Senior Systems Engineer': 'Senior Staff',
    'Principal Software Engineer': 'Technique Leader'
}

df['title'] = df['title'].replace(title_mapping)

df

title  salary
0            Engineer  135575
1     Senior Engineer  195208
2  Assistant Engineer   98554
3               Staff  123232
4        Senior Staff  144377
5    Technique Leader  169861

Calcular para cada empleado el ratio entre su salario y el salario de referencia en 2003, teniendo en cuenta la inflación media desde el año 2003 (3.5%).

In [ ]:
%%sql

WITH inflation_adjusted_salaries AS (
    SELECT
        title,
        ROUND(salary / POWER(1.035, 20), 2) AS adjusted_salary_2003
    FROM (VALUES
        ('Engineer', 135575),
        ('Senior Engineer', 195208),
        ('Assistant Engineer', 98554),
        ('Staff', 123232),
        ('Senior Staff', 144377),
        ('Technique Leader', 169861)
    ) AS reference_salaries(title, salary)
)
, employee_ratios AS (
    SELECT
        e.emp_no,
        e.first_name || ' ' || e.last_name AS employee_name,
        t.title,
        s.amount AS salary_2003,
        ias.adjusted_salary_2003,
        ROUND(s.amount / NULLIF(ias.adjusted_salary_2003, 0), 2) AS ratio
    FROM employee e
    JOIN salary s ON e.emp_no = s.emp_no AND s.from_date <= '2003-12-31' AND s.to_date > '2003-12-31'
    JOIN title t ON e.emp_no = t.emp_no AND t.from_date <= '2003-12-31' AND t.to_date > '2003-12-31'
    JOIN inflation_adjusted_salaries ias ON t.title = ias.title
)

SELECT * FROM employee_ratios
ORDER BY ratio DESC;



 * postgresql+psycopg2://postgres:***@5.tcp.eu.ngrok.io:16267/postgres


emp_no         employee_name               title  salary_2003  \
0     10024        Suzette Pettey  Assistant Engineer        96646   
1     10603         Wayne Restivo               Staff       118613   
2     10897        Arno Kumaresan        Senior Staff       136004   
3     10539        Yucel Reinhard        Senior Staff       126614   
4     10548  Ramalingam Gunderson        Senior Staff       126034   
..      ...                   ...                 ...          ...   
797   10343        Dinah Syrzycki     Senior Engineer        47242   
798   10786      Patricia dAstous     Senior Engineer        47349   
799   10506        Kazuhisa Kroll     Senior Engineer        46387   
800   10027        Divier Reistad     Senior Engineer        46145   
801   10003         Parto Bamford     Senior Engineer        43311   

    adjusted_salary_2003 ratio  
0               49529.88  1.95  
1               61932.20  1.92  
2               72558.95  1.87  
3               72558.95  1.74  
4               72558.95  1.74  
..                   ...   ...  
797             98104.88  0.48  
798             98104.88  0.48  
799             98104.88  0.47  
800             98104.88  0.47  
801             98104.88  0.44  

[802 rows x 6 columns]

Lo agrupamos por rol, para que sea más visual,

In [ ]:
%%sql
WITH inflation_adjusted_salaries AS (
    SELECT
        title,
        salary / POWER(1.035, 20) AS adjusted_salary_2003
    FROM (VALUES
        ('Engineer', 135575),
        ('Senior Engineer', 195208),
        ('Assistant Engineer', 98554),
        ('Staff', 123232),
        ('Senior Staff', 144377),
        ('Technique Leader', 169861)
    ) AS reference_salaries(title, salary)
)

, employee_ratios AS (
    SELECT
        t.title,
        s.amount AS salary_2003,
        ias.adjusted_salary_2003,
        s.amount / NULLIF(ias.adjusted_salary_2003, 0) AS ratio
    FROM employee e
    JOIN salary s ON e.emp_no = s.emp_no AND s.from_date <= '2003-12-31' AND s.to_date > '2003-12-31'
    JOIN title t ON e.emp_no = t.emp_no AND t.from_date <= '2003-12-31' AND t.to_date > '2003-12-31'
    JOIN inflation_adjusted_salaries ias ON t.title = ias.title
)

SELECT
    title,
    ROUND(AVG(salary_2003), 2) AS average_salary_2003,
    ROUND(AVG(ratio), 2) AS average_ratio
FROM employee_ratios
GROUP BY title
ORDER BY average_ratio DESC;


 * postgresql+psycopg2://postgres:***@5.tcp.eu.ngrok.io:16267/postgres


title average_salary_2003 average_ratio
0  Assistant Engineer            59964.75          1.21
1        Senior Staff            78954.60          1.09
2               Staff            66892.28          1.08
3            Engineer            61082.29          0.90
4    Technique Leader            68578.69          0.80
5     Senior Engineer            71987.70          0.73

Media del ratio:

In [ ]:
average_ratio = [1.21, 1.09, 1.08, 0.90, 0.80, 0.73]
media = sum(average_ratio) / len(average_ratio)
print(round(media, 2))

0.97


**Conclusiones**:
El ratio que hemos calculado representa la relación entre el salario que un empleado recibió en 2003 y el salario de referencia para su puesto en ese mismo año, después de ajustar la referencia de 2023 por inflación.

- El ratio 0,97 nos indica que la empresa UAX Tech en el año 2003 de media pagaba bastante acorde al sueldo promedio del 2003.

- Si nos fijamos el ratio promedio por rol, vemos que la empresa pagaba un 21% más en el rol de Assistant Engineer respecto al promedio del mercado laboral del año 2003, lo que quiere decir que la empresa ha perdido dinero ya que ha pagado de más en ese rol. Esto explica porque la empresa excedió en gasto salarial y quebro a causa de ello.

- Pasa lo mismo en el rol de Senior Staff y Staff.

Ha habido un sobregasto en sueldos, lo que sugiere que hay que ajustar al promedio del mercado laboral los sueldos de los roles Assistant Engineer, Senior Staff y Staff.


### 5. Estrategia de Reestructuración

Antes de planificar la estrategia de reestructuración necesitamos tener el contexto de la situación y definir el objetivo de la estrategia:

In [ ]:
%%sql
WITH ActiveSalaries AS (
    -- Identificamos los salarios de los empleados activos el 31-12-2002
    SELECT
        s.emp_no,
        s.amount AS salario
    FROM salary s
    WHERE s.from_date <= '2002-12-31' AND s.to_date > '2002-12-31'
)

SELECT
    SUM(salario) AS TotalSalariesExpense
FROM ActiveSalaries;


 * postgresql+psycopg2://postgres:***@5.tcp.eu.ngrok.io:16267/postgres


totalsalariesexpense
0              57577804

El problema principal es que la empresa gasta en salarios aproximadamente un 10% más a la facturación. Por lo tanto el **objetivo** es reducir en un 15% el gasto en salarios, que ha fecha de 31-12-2002 era de 57.577.804 dolares, lo cual significa que hay que recortar 8.636.670,6$)

De este modo daremos un margen del 5% (lo conseguimos reduciendo el gasto en sueldos en un 15%) para conseguir solvencia y liquidez, dando lugar a una mejor situación en las cuentas.

Teniendo en cuenta lo analizado, las medidas para la reestructuración que se llevará a cabo con el proposito de mejorar las cuentas de UAX Tech es la siguiente:


1. Se prejubilará a los empleados con más antigüedad y que estén cerca de la edad de jubilación. Con esto minimizamos el impacto moral de los empleados veteranos

In [ ]:
%%sql
WITH EligibleForPreRetirement AS (
    SELECT
        emp_no,
        salarios,
        edad,
        antiguedad
    FROM (
        SELECT
            e.emp_no,
            s.amount AS salarios,
            EXTRACT(YEAR FROM age('2002-12-31', e.birth_date))::INT AS edad,
            EXTRACT(YEAR FROM age('2002-12-31', e.hire_date))::INT AS antiguedad
        FROM employee e
        JOIN salary s ON e.emp_no = s.emp_no
        WHERE s.from_date <= '2002-12-31' AND s.to_date > '2002-12-31'
    ) AS DerivedTable
    WHERE edad >= 50 AND antiguedad >= 15
)

SELECT
    COUNT(emp_no) AS Number_Of_Employees_To_PreRetire,
    SUM(salarios) AS Total_Savings
FROM EligibleForPreRetirement;



 * postgresql+psycopg2://postgres:***@5.tcp.eu.ngrok.io:16267/postgres


number_of_employees_to_preretire  total_savings
0                                15        1200885

2. Se realizará una reducción salarial selectiva del 15%, basado en los roles y departamentos que tengan salarios promedios más altos.

In [ ]:
%%sql
WITH AverageSalaries AS (
    SELECT
        t.title AS rol,
        d.dept_name AS departamento,
        AVG(s.amount) AS average_salary
    FROM salary s
    JOIN title t ON s.emp_no = t.emp_no AND s.to_date > '2002-12-31' AND t.to_date > '2002-12-31'
    JOIN dept_emp de ON s.emp_no = de.emp_no AND de.to_date > '2002-12-31'
    JOIN department d ON de.dept_no = d.dept_no
    GROUP BY t.title, d.dept_name
),
TopSalaries AS (
    SELECT
        rol,
        departamento,
        average_salary
    FROM AverageSalaries
    WHERE average_salary > (SELECT percentile_cont(0.75) WITHIN GROUP (ORDER BY average_salary) FROM AverageSalaries)
)

-- Ahora calculamos el ahorro total basado en una reducción del 10% (por ejemplo) para estos roles y departamentos.

, Savings AS (
    SELECT
        ts.rol,
        ts.departamento,
        SUM(s.amount) * 0.15 AS potential_savings -- asumimos una reducción del 10%
    FROM TopSalaries ts
    JOIN title t ON ts.rol = t.title
    JOIN dept_emp de ON t.emp_no = de.emp_no AND de.to_date > '2002-12-31'
    JOIN department d ON de.dept_no = d.dept_no AND ts.departamento = d.dept_name
    JOIN salary s ON t.emp_no = s.emp_no AND s.to_date > '2002-12-31'
    GROUP BY ts.rol, ts.departamento
)

SELECT
    rol,
    departamento,
    potential_savings
FROM Savings;


 * postgresql+psycopg2://postgres:***@5.tcp.eu.ngrok.io:16267/postgres


rol departamento potential_savings
0   Senior Engineer  Development        1306241.10
1   Senior Engineer   Production        1241091.90
2      Senior Staff      Finance         325421.70
3      Senior Staff    Marketing         347544.75
4      Senior Staff        Sales        1294168.80
5             Staff    Marketing         319616.10
6             Staff        Sales        1241319.90
7  Technique Leader     Research          35667.75

3. Realizamos una reducción de salario del 21% a todos los empleados con el rol de Assistant Engineer, ya que reciben un salario superior del 21% más respecto a lo que ofrece de media el mercado laboral de Estados Unidos.

In [ ]:
%%sql
-- Paso 1: Calculamos la reducción de salario para todos los empleados con el rol de "Assistant Engineer"
WITH salary_reduction AS (
    SELECT
        s.emp_no,
        s.amount,
        s.amount * 0.21 AS reduction_amount,
        s.amount * 0.79 AS new_salary -- 100% - 21% = 79%
    FROM salary s
    JOIN title t ON s.emp_no = t.emp_no
    WHERE t.title = 'Assistant Engineer'
    AND s.from_date <= '2002-12-31' AND s.to_date > '2002-12-31'
    AND t.from_date <= '2002-12-31' AND t.to_date > '2002-12-31'
)

-- Paso 2: Sumamos el total de ahorro que supone la reducción de salario
SELECT
    SUM(reduction_amount) AS total_savings
FROM salary_reduction;



 * postgresql+psycopg2://postgres:***@5.tcp.eu.ngrok.io:16267/postgres


total_savings
0     151111.17

**CONCLUSIÓN**: El gasto salarial ahorrado total tras aplicar estas medidas calculados sobre los datos de la empresa a fecha de 31-12-2002, es de: 7463068.17

Por lo tanto no hemos conseguido el objetivo planeado pero si hemos reducido aproximadamente en un 13% los gastos en salarios (12.96%), lo que supone un plan de reestructuración viable evitando despidos masivos.

In [ ]:
ahorro_prejubilaciones = [1200885]
ahorro_reduccion_salarial = [1306241.10, 1241091.90,	325421.70, 347544.75, 1294168.80, 319616.10, 1241319.90, 35667.75]
ahorro_salario_assistant_engineer = [151111.17]

total_ahorro = sum(ahorro_prejubilaciones) + sum(ahorro_reduccion_salarial) + sum(ahorro_salario_assistant_engineer)
print(f"El ahorro total es de: {total_ahorro}")

El ahorro total es de: 7463068.17


In [ ]:
 #Guradamos ne html:
 %%shell
 jupyter nbconvert --to html "/content/drive/MyDrive/Colab Notebooks/Fundamento_de_la_Ciencia_de_Datos/Examen 1/2_Examen_Isaac_Mohamed.ipynb"

[NbConvertApp] Converting notebook /content/drive/MyDrive/Colab Notebooks/Fundamento_de_la_Ciencia_de_Datos/Examen 1/2_Examen_Isaac_Mohamed.ipynb to html
[NbConvertApp] Writing 865829 bytes to /content/drive/MyDrive/Colab Notebooks/Fundamento_de_la_Ciencia_de_Datos/Examen 1/2_Examen_Isaac_Mohamed.html
